### PR TITLE
perf(workbench): accelerate regex log filtering with Roaring Bitmap TrigramIndex and optimized SearchIndex

### PR DESCRIPTION
**Pull request by kyasbal**
## Summary

This PR optimizes log filtering performance in the KHI Workbench by introducing a precomputed in-memory `SearchIndex` and accelerating regular expression evaluation using a Roaring Bitmap-based `TrigramIndex`. It also adds real-time search index construction progress streaming via Connect-RPC and optimizes memory consumption during workbench ingestion.

## Background & Motivation

1. **High CEL Log Filtering Latency on Large Datasets**: Evaluating regex filters (e.g., `has(log.body.message) && log.body.message.matches("...")`) across hundreds of thousands of logs previously required full RE2 regex evaluations on serialized YAML strings for every single log entry.
2. **Sub-string Pruning with Roaring Bitmaps**: Most regex patterns specify fixed literal substrings (e.g., `pod-name.*failed`). By indexing 3-character n-grams (trigrams) of interned YAML strings into compressed bitsets (`roaring.Bitmap`), candidate structs/logs can be filtered with high precision in microseconds before invoking RE2.
3. **Memory Footprint Optimization**: Precomputing full path maps and log ID lists for every timeline previously created significant heap overhead. Storing `LogData` directly in a flat slice and resolving paths and log IDs on demand keeps memory consumption minimal while retaining O(1) log lookups.
4. **User Experience During Ingestion**: Constructing the search index and trigram index can take several seconds for massive files. A dedicated streaming RPC (`StreamIndexProgress`) and frontend progress UI provide clear feedback to the user.

## API Endpoint Path Changes

| Change | Protocol / Type | HTTP Method | Endpoint Path | Description |
| :--- | :--- | :--- | :--- | :--- |
| ➕ **Added** | Connect-RPC (Streaming) | `POST` | `/api.v1.WorkbenchService/StreamIndexProgress` | Streams search index construction progress percentages and lifecycle states (`BUILDING`, `READY`, `FAILED`). |

### 1. Protobuf (`proto/api/v1/workbench.proto`)
- Defined `StreamIndexProgress` RPC returning server-streaming `StreamIndexProgressResponse` with `IndexState` (`INDEX_STATE_BUILDING`, `INDEX_STATE_READY`, `INDEX_STATE_FAILED`).
- Generated Go and TypeScript Connect stubs.

### 2. Go Backend
- **`pkg/common/worker`**:
  - Implemented `ParallelChunkMap` supporting concurrent chunk processing, fine-grained progress reporting, error aggregation, and context cancellation.
- **`pkg/server/workbench` (`index.go`, `reader.go`, `workbench.go`)**:
  - Introduced `SearchIndex` precomputing timeline hierarchies, log severities, and interned string references.
  - Flattened `LogData` into `[]cel.LogData` with O(1) index access (`GetLog`), eliminating `LogMap` pointer overhead.
  - Replaced static timeline `LogIDs` and `Path` maps with on-demand traversal (`ForEachLogID`, `ComputePath`).
  - Implemented `StreamIndexProgress` Connect-RPC handler streaming index generation progress.
  - Released raw `logChunks` and `timelineChunks` after search index construction to reclaim heap memory.
- **`pkg/server/workbench/cel` (`trigram.go`, `trigram_query.go`, `functions.go`, `env.go`)**:
  - Implemented `TrigramIndex` using Roaring Bitmaps (`github.com/RoaringBitmap/roaring`) for fast substring candidate pruning.
  - Implemented 2-stage Boolean AST (`TrigramQuery`) with AST simplification (AND, OR, NOT pruning) from regex patterns.
  - Accelerated CEL `matches` and `contains` functions by executing trigram candidate pruning before RE2 regex matching.

### 3. Angular Frontend
- **`src/app/index-progress`**:
  - Implemented `IndexProgressSmartComponent` and `IndexProgressLayoutComponent` following the Smart-Dumb component pattern.
  - Added Storybook stories (`index-progress.stories.ts`) and unit tests (`index-progress-smart.component.spec.ts`, `index-progress-layout.component.spec.ts`).
- **`src/app/services/api/workbench`**:
  - Added `StreamIndexProgress` client method in `WorkbenchClientService`.
- **`src/app/pages/main`**:
  - Integrated `khi-index-progress-smart` into the main application layout.

## Verification

- **Backend Unit Tests**: `make test-go` passed across all packages.
- **Frontend Unit Tests**: `make test-web` passed (760/760 tests).
- **Linters & Formatters**: `make lint-go`, `make lint-web`, `make lint-md`, `make format-go`, `make format-web`, and `make pre-commit` passed with 0 issues.

### CODE REVIEW

File: pkg/common/worker/parallel.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+			<-doneCh
+			pct, msg := formatProgress(processedCount.Load())
+			_ = onProgress(pct, msg)
+		}
+	} else {
+		stopReporter = func() {}
+	}
+
+	results := make([]R, numWorkers)
+	chunkSize := (len(items) + numWorkers - 1) / numWorkers
```
[the remark this answers is not in the document]
Comment (kyasbal) [reply pkg/common/worker/parallel.go]: Done. Adjusted numWorkers calculation when chunkSize rounding yields fewer chunks, and added unit test coverage.

File: pkg/server/workbench/filter_log_cel.go [hunk 2]
```diff
@@ -49,12 +53,27 @@ func (f *LogCELFilter) Process(
...
+		if err := pprof.StartCPUProfile(pprofFile); err == nil {
+			defer func() {
+				pprof.StopCPUProfile()
+				_ = pprofFile.Close()
+				slog.InfoContext(ctx, "saved CPU profile for log CEL filter", "path", pprofPath)
+			}()
+		} else {
+			_ = pprofFile.Close()
+		}
+	}
```
[the remark this answers is not in the document]
Comment (kyasbal) [reply pkg/server/workbench/filter_log_cel.go]: Done. Removed debug CPU profiling code and unused imports from filter_log_cel.go.

File: pkg/server/workbench/index.go [hunk 3]
```diff
@@ -15,293 +15,420 @@
...
+
+	if err := onProgress(apiv1.OpenWorkbenchResponse_STAGE_INDEXING_DATA, 100.0, "Text search index complete."); err != nil {
+		return nil, err
+	}
+
+	return trigramIndex, nil
+}
+
+// BuildAsyncIndexesWithProgress populates asynchronous search indexes (such as struct YAMLs and trigram indexes) on the target SearchIndex while streaming progress updates.
+func (w *Workbench) BuildAsyncIndexesWithProgress(targetIndex *SearchIndex, onProgress ProgressCallback) error {
```
[the remark this answers is not in the document]
Comment (kyasbal) [reply pkg/server/workbench/index.go]: Done. Propagated ctx context.Context through BuildAsyncIndexesWithProgress, BuildStructYAMLIndexWithProgress, BuildTrigramIndexWithProgress, and BuildFromStructYAMLs down to worker.ParallelChunkMap.

File: web/src/app/index-progress/index-progress-smart.component.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            this.isVisibleSignal.set(false);
+            this.dismissTimeout = null;
+          }, 1500);
+        }
+      } else {
+        if (!isReady) {
+          this.isVisibleSignal.set(false);
+        }
+      }
+    });
```
[the remark this answers is not in the document]
Comment (kyasbal) [reply web/src/app/index-progress/index-progress-smart.component.ts]: Done. Added allowSignalWrites: true to effect in IndexProgressSmartComponent.

File: web/src/app/services/api/workbench/workbench-client.service.ts [hunk 5]
```diff
@@ -82,6 +83,13 @@ export class WorkbenchClientService implements OnDestroy {
     WorkbenchClientService.STRUCT_YAML_CACHE_CAPACITY,
   );
 
+  private readonly indexStateSignal =
+    signal<StreamIndexProgressResponse_IndexState>(
+      StreamIndexProgressResponse_IndexState.UNSPECIFIED,
+    );
+  private readonly indexProgressPercentageSignal = signal<number>(0);
+  private readonly indexMessageSignal = signal<string>('');
```
[the remark this answers is not in the document]
Comment (kyasbal) [reply web/src/app/services/api/workbench/workbench-client.service.ts:91]: Done. Added indexProgressAbortController field to WorkbenchClientService.

File: web/src/app/services/api/workbench/workbench-client.service.ts [hunk 6]
```diff
@@ -148,11 +189,42 @@ export class WorkbenchClientService implements OnDestroy {
       this.structYamlCache.clear();
       this.activeWorkbenchIdSignal.set(workbenchId);
       this.startHeartbeat(workbenchId);
+      void this.streamIndexProgress(workbenchId);
```
[the remark this answers is not in the document]
Comment (kyasbal) [reply web/src/app/services/api/workbench/workbench-client.service.ts]: Done. Aborted any active index progress stream before starting a new stream in openWorkbench.

File: web/src/app/services/api/workbench/workbench-client.service.ts [hunk 7]
```diff
@@ -176,6 +248,11 @@ export class WorkbenchClientService implements OnDestroy {
     this.stopHeartbeat();
     this.structYamlCache.clear();
     this.activeWorkbenchIdSignal.set(null);
+    this.indexStateSignal.set(
+      StreamIndexProgressResponse_IndexState.UNSPECIFIED,
+    );
+    this.indexProgressPercentageSignal.set(0);
+    this.indexMessageSignal.set('');
```
[the remark this answers is not in the document]
Comment (kyasbal) [reply web/src/app/services/api/workbench/workbench-client.service.ts:286]: Done. Aborted active index progress stream and reset abort controller in closeWorkbench.

File: proto/api/v1/workbench.proto [hunk 8]
```diff
@@ -25,6 +25,9 @@ service WorkbenchService {
   // Opens or attaches to an in-memory Workbench session with real-time loading progress streaming.
   rpc OpenWorkbench(OpenWorkbenchRequest) returns (stream OpenWorkbenchResponse);
 
+  // Streams the search index construction progress and final status for an active Workbench session.
+  rpc StreamIndexProgress(StreamIndexProgressRequest) returns (stream StreamIndexProgressResponse);
```
Comment (kyasbal) [inline proto/api/v1/workbench.proto]: For consistency with the other Watch requests, this should be named as WatchIndexProgress and reconnect every 30 sec. Not all the environment supports the long response time.

File: pkg/server/workbench/cel/trigram_query_test.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+		{
+			name:       "nested captures and plus operators",
+			pattern:    "((pod-a)+)",
+			wantString: `AND(TERM("pod"), TERM("od-"), TERM("d-a"))`,
+		},
+		{
+			name:       "only wildcards becomes ALL",
+			pattern:    ".*",
+			wantString: "ALL",
+		},
```
Comment (kyasbal) [inline pkg/server/workbench/cel/trigram_query_test.go:99]: Add tests for non unicode string.

File: pkg/server/workbench/cel/trigram_test.go [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+    image: nginx:latest
+`
+	yaml2 := `metadata:
+  name: pod-b
+  namespace: kube-system
+spec:
+  containers:
+  - name: coredns
+    image: coredns:v1.9
+`
```
Comment (kyasbal) [inline pkg/server/workbench/cel/trigram_test.go:48]: Add tests for no unicode search.

File: proto/api/v1/workbench.proto (hunk 8 above)
Comment (kyasbal) [reply proto/api/v1/workbench.proto]: Done. Renamed RPC and message types to `WatchIndexProgress`, updated server to terminate the stream every 30s to prevent proxy timeouts, and updated the frontend client to reconnect until the index reaches a terminal state (`READY` or `FAILED`) or is aborted.

File: pkg/server/workbench/cel/trigram_query_test.go (hunk 9 above)
Comment (kyasbal) [reply pkg/server/workbench/cel/trigram_query_test.go:99]: Done. Added test cases for multi-byte Unicode strings (Kanji/Katakana), short multi-byte strings (<3 runes), mixed ASCII/Unicode strings, and Unicode case folding with accents in `trigram_query_test.go`.

File: pkg/server/workbench/cel/trigram_test.go (hunk 10 above)
Comment (kyasbal) [reply pkg/server/workbench/cel/trigram_test.go:48]: Done. Added `TestTrigramIndex_Unicode` in `trigram_test.go` verifying multi-byte Unicode indexing, candidate filtering, mixed ASCII/Unicode queries, and short rune fallback against documents with non-ASCII text.
